pkg/cli/api_response_typed.go

```diff
@@ -0,0 +1,85 @@
+package cli
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/tidwall/pretty"
+	"io"
+	"net/http"
+	"strings"
+)
+
+var _ APIResponse = &TypedAPIResponse[struct{}]{}
+
+type TypedAPIResponse[TBody any] struct {
+	StatusCode  int   `json:"statusCode"`
+	Body        TBody `json:"body"`
+	Error       error `json:"error"`
+	contentType string
+}
+
+func NewTypedAPIResponse[TBody any](body TBody) func(resp *http.Response, err error) *TypedAPIResponse[TBody] {
+	return func(resp *http.Response, err error) *TypedAPIResponse[TBody] {
+		apiRes := TypedAPIResponse[TBody]{
+			Error: err,
+		}
+		if resp == nil {
+			return &apiRes
+		}
+
+		apiRes.StatusCode = resp.StatusCode
+		apiRes.contentType = resp.Header.Get("Content-Type")
+
+		out, err := io.ReadAll(resp.Body)
+		if err != nil {
+			apiRes.Error = fmt.Errorf("failed to parse body: %s", err.Error())
+			return &apiRes
+		}
+
+		switch strings.Split(resp.Header.Get("Content-Type"), ";")[0] {
+		case "application/json":
+			if err := json.Unmarshal(out, &body); err != nil {
+				apiRes.Error = errors.Wrapf(err, "failed to parse body as JSON")
+				return &apiRes
+			}
+		case "text/plain":
+			apiRes.Error = fmt.Errorf(strings.TrimSpace(string(out)))
+			return &apiRes
+		default:
+			apiRes.Error = fmt.Errorf("unknown content type %s", resp.Header.Get("Content-Type"))
```

Comment:
We should error with the split string we are testing against:
```suggestion
			apiRes.Error = fmt.Errorf("unknown content type %s", strings.Split(resp.Header.Get("Content-Type"), ";")[0])
```